hoverkite-protocol/src/lib.rs

```diff
@@ -87,87 +87,89 @@ impl Command {
-        let first = buf.get(0).ok_or(WouldBlock)?;
-        match first {
-            b'l' => Ok(Command::SetSideLed(ascii_to_bool(
-                *buf.get(1).ok_or(WouldBlock)?,
-            )?)),
-            b'o' => Ok(Command::SetOrangeLed(ascii_to_bool(
-                *buf.get(1).ok_or(WouldBlock)?,
-            )?)),
-            b'r' => Ok(Command::SetRedLed(ascii_to_bool(
-                *buf.get(1).ok_or(WouldBlock)?,
-            )?)),
-            b'g' => Ok(Command::SetGreenLed(ascii_to_bool(
-                *buf.get(1).ok_or(WouldBlock)?,
-            )?)),
-            b'b' => Ok(Self::ReportBattery),
-            b'c' => Ok(Self::ReportCharger),
-            b'S' => {
-                if buf.len() < 5 {
+        let command = match *buf {
+            [] | [b'l'] | [b'o'] | [b'r'] | [b'g'] => return Err(WouldBlock),
+            [b'l', on] => Self::SetSideLed(ascii_to_bool(on)?),
+            [b'o', on] => Self::SetOrangeLed(ascii_to_bool(on)?),
+            [b'r', on] => Self::SetRedLed(ascii_to_bool(on)?),
+            [b'g', on] => Self::SetGreenLed(ascii_to_bool(on)?),
+            [b'b'] => Self::ReportBattery,
+            [b'c'] => Self::ReportCharger,
+            [b'S', ref rest @ ..] => {
+                if rest.len() < 4 {
                     return Err(WouldBlock);
+                } else if let [min_lsb, min_msb, max_lsb, max_msb] = *rest {
+                    let min_power = i16::from_le_bytes([min_lsb, min_msb]);
+                    let max_power = i16::from_le_bytes([max_lsb, max_msb]);
+                    Self::SetMaxSpeed(min_power..=max_power)
+                } else {
+                    return Err(Other(ParseError));
                 }
-                let min_power = i16::from_le_bytes(buf[1..3].try_into().unwrap());
-                let max_power = i16::from_le_bytes(buf[3..5].try_into().unwrap());
-
-                Ok(Self::SetMaxSpeed(min_power..=max_power))
             }
-            b'K' => {
-                if buf.len() < 3 {
+            [b'K', ref rest @ ..] => {
+                if rest.len() < 2 {
                     return Err(WouldBlock);
+                } else if let [lsb, msb] = *rest {
+                    let spring = u16::from_le_bytes([lsb, msb]);
+                    Self::SetSpringConstant(spring)
+                } else {
+                    return Err(Other(ParseError));
                 }
-                let spring = u16::from_le_bytes(buf[1..3].try_into().unwrap()).into();
-                Ok(Self::SetSpringConstant(spring))
             }
-            b'n' => Ok(Self::RemoveTarget),
-            b'T' => {
-                if buf.len() < 9 {
+            [b'n'] => Self::RemoveTarget,
+            [b'T', ref rest @ ..] => {
+                if rest.len() < 8 {
                     return Err(WouldBlock);
+                } else if let [b0, b1, b2, b3, b4, b5, b6, b7] = *rest {
+                    let target = i64::from_le_bytes([b0, b1, b2, b3, b4, b5, b6, b7]);
```

Comment:
Splitting up to a bunch of named variables just to immediately put them back into an array again looks a bit weird. Why not just use a slice like before?

Reply:
fair. It's annoying that you can't unpack a slice into an array like `[b'T', ref rest @ [u8; 8]] => {}` or something